[15.0][MIG] account_spread_cost_revenue

Saran440 opened the pull request:
This module migration from standard and add PR enhance https://github.com/OCA/account-financial-tools/pull/1580

astirpe committed:
[ADD] account_spread_cost_revenue

astirpe committed:
[MIG] account_spread_cost_revenue to V12

astirpe committed:
Improvements as per feedback

astirpe committed:
Add button to post all spread lines

astirpe committed:
Improvements as per feedback

astirpe committed:
flake8

astirpe committed:
Better tests

astirpe committed:
Better tests

astirpe committed:
Better tests

astirpe committed:
Handle negative amounts

astirpe committed:
Fix creation of credit invoices

astirpe committed:
Add Dutch translation

astirpe committed:
Update Copyright dates

astirpe committed:
Add option to disable spread planning + code review

astirpe committed:
Add option to force the posting of accounting move

astirpe committed:
Update nl.po

astirpe committed:
Coverage

astirpe committed:
Code review + add coverage

astirpe committed:
Add partner in move lines

astirpe committed:
Add option to auto-archive spreads

oca-travis committed:
[UPD] Update account_spread_cost_revenue.pot

OCA-git-bot committed:
[UPD] README.rst

Martronic-SA committed:
Added translation using Weblate (French)

Martronic-SA committed:
Translated using Weblate (French)

Currently translated at 90.0% (135 of 150 strings)

Translation: account-financial-tools-12.0/account-financial-tools-12.0-account_spread_cost_revenue
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-12-0/account-financial-tools-12-0-account_spread_cost_revenue/fr/

Martronic-SA committed:
Translate wizard choices.

Extra wizard selection ar not translated.

oca-travis committed:
[UPD] Update account_spread_cost_revenue.pot

OCA-git-bot committed:
account_spread_cost_revenue 12.0.1.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-financial-tools-12.0/account-financial-tools-12.0-account_spread_cost_revenue
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-12-0/account-financial-tools-12-0-account_spread_cost_revenue/

Martronic-SA committed:
Translated using Weblate (French)

Currently translated at 90.2% (138 of 153 strings)

Translation: account-financial-tools-12.0/account-financial-tools-12.0-account_spread_cost_revenue
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-12-0/account-financial-tools-12-0-account_spread_cost_revenue/fr/

alvarorib committed:
Added translation using Weblate (Portuguese)

alvarorib committed:
Translated using Weblate (Portuguese)

Currently translated at 35.3% (54 of 153 strings)

Translation: account-financial-tools-12.0/account-financial-tools-12.0-account_spread_cost_revenue
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-12-0/account-financial-tools-12-0-account_spread_cost_revenue/pt/

kittiu committed:
[12.0][ENH] account_spreaad_cost_revenue, add optional expense/revenue account

oca-travis committed:
[UPD] Update account_spread_cost_revenue.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
account_spread_cost_revenue 12.0.1.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-financial-tools-12.0/account-financial-tools-12.0-account_spread_cost_revenue
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-12-0/account-financial-tools-12-0-account_spread_cost_revenue/

kittiu committed:
[12.0][IMP] account_spread_cost_revenue, auto spread on validate invoice

espo-tony committed:
[12.0][FIX] account_spread_cost_revenue: convert total amount to spread currency

OCA-git-bot committed:
account_spread_cost_revenue 12.0.1.1.1

OCA-git-bot committed:
account_spread_cost_revenue 12.0.1.2.0

astirpe committed:
pre-commit

astirpe committed:
[13.0][MIG] account_spread_cost_revenue

kittiu committed:
[IMP] account_spread_cost_revenue: black, isort, prettier

kittiu committed:
[14.0][MIG] account_spread_cost_revenue

Commit:
[UPD] Update account_spread_cost_revenue.pot

OCA-git-bot committed:
[UPD] README.rst

ps-tubtim committed:
[ENH] account_spread_cost_revenue: add calculate by days feature

Saran440 committed:
[IMP] account_spread_cost_revenue: black, isort, prettier

Saran440 committed:
[15.0][MIG] account_spread_cost_revenue

astirpe commented on account_spread_cost_revenue/models/account_move_line.py:
```diff
@@ -97,7 +97,9 @@ def write(self, vals):
         reconciled_lines = self.filtered(lambda l: l.reconciled)
-        msg_line = _("Move line: %s (%s), account code: %s\n")
+        msg_line = _(
+            "Move line: %(line_id)s (%(line_name)s), account code: %(account_code)s\n"
```
Why changing this line? This causes a failure when executing this line:  `msg += msg_line % (line.id, line.name, line.account_id.code)`

BT-anieto replied:
Agree with @astirpe. Could you check it please @Saran440 ?

BT-anieto commented on account_spread_cost_revenue/models/account_move_line.py:
```diff
@@ -146,10 +148,9 @@ def _filter_line(aline, iline):
-                        "invoice line, %s"
-                    )
-                    % (len(template), line.display_name)
+                        "Too many auto spread templates ({}) matched with the "
+                        "invoice line, {}"
+                    ).format(len(template), line.display_name)
```
Do not change the translations. Format is not recommended.

BT-anieto commented on account_spread_cost_revenue/models/account_spread.py:
```diff
@@ -520,7 +520,7 @@ def _message_post_unlink_invoice_line(self):
                 "data-oe-id=%d>%s</a>" % (spread.invoice_id.id, _("Invoice"))
             )
-            msg_body = _("Unlinked invoice line '%s' (view %s).") % (
+            msg_body = _("Unlinked invoice line '{}' (view {}).").format(
```
Do not change the translations. Format is not recommended

BT-anieto commented on account_spread_cost_revenue/models/account_spread.py:
```diff
@@ -529,7 +529,7 @@ def _message_post_unlink_invoice_line(self):
                 "data-oe-id=%d>%s</a>" % (spread.id, _("Spread"))
             )
-            msg_body = _("Unlinked '%s' (invoice line %s).") % (
+            msg_body = _("Unlinked '{}' (invoice line {}).").format(
```
Do not change the translations. Format is not recommended

BT-anieto commented on account_spread_cost_revenue/models/account_spread_template.py:
```diff
@@ -169,7 +169,9 @@ def _check_auto_spread_ids_unique(self):
         if results:
             raise UserError(
-                _("Followings are duplicated combinations,\n\n%s" % "\n".join(results))
+                _("Followings are duplicated combinations,\n\n{}").format(
```
Do not change the translations. Format is not recommended

BT-anieto reviewed (changes requested):
Thx for the PR! Try to reduce the changes at maximum when it is a migration. Moreover, format is not recommended for odoo translations. Could you chan ge it?

On the other hand, ensure that you squash the commits following https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

Saran440 commented:
@BT-anieto @astirpe Hi, Thank you for your review
I change it back to %(name)s because %s not support pre-commit it will error

```
account_spread_cost_revenue/models/account_move_line.py:100: [W8120(translation-positional-used), AccountMoveLine._check_spread_reconcile_validity] Translation method _('Move line: %s (%s), account code: %s\n') is using positional string printf formatting. Use named placeholder `_("%(placeholder)s")` instead.

account_spread_cost_revenue/models/account_move_line.py:148: [W8120(translation-positional-used), AccountMoveLine.create_auto_spread] Translation method _('Too many auto spread templates (%s) matched with the invoice line, %s') is using positional string printf formatting. Use named placeholder `_("%(placeholder)s")` instead.

account_spread_cost_revenue/models/account_spread.py:523: [W8120(translation-positional-used), AccountSpread._message_post_unlink_invoice_line] Translation method _("Unlinked invoice line '%s' (view %s).") is using positional string printf formatting. Use named placeholder `_("%(placeholder)s")` instead.

account_spread_cost_revenue/models/account_spread.py:532: [W8120(translation-positional-used), AccountSpread._message_post_unlink_invoice_line] Translation method _("Unlinked '%s' (invoice line %s).") is using positional string printf formatting. Use named placeholder `_("%(placeholder)s")` instead.

```

If you ok for this commit 33be08740bc71c36fdf420dc8ab312c2c749c788 i will merge this commit into 1 commit
What do you think?

BT-anieto commented:
Hi @Saran440 , you still need to follow https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests
You need to squash the commits and you can only push 2 commits: one of the refactor and another one with the code changes.

Saran440 committed:
[FIX] account_spread_cost_revenue: Skip creating a template

Bug fix: Duplicate spread creation when clicking 'Create Move'

Error Steps:
1. Create a spread template.
2. Create a vendor bill and select the same account as the template, then confirm it.
3. Go to 'Spread Costs/Revenues' and click 'create move' on line.
4. It creates a new move and a new template. (It should not create a new template)

Saran440 commented:
I found a bug in module account_spread_cost_revenue https://github.com/OCA/account-financial-tools/pull/1646/commits/e317f12a366268be8a8ad9b3c7c627663acb8256
Should we fixed it in this PR or next PR?

Step to error:
1. create spread templates
![Selection_001](https://github.com/OCA/account-financial-tools/assets/20896369/b0482823-5cc5-43d0-b1cf-aef0abbbf725)

2. create vendor bills and select the same account as the template > confirm
3. Go to 'Spread Costs/Revenues' > select spread cost that create from template > click 'create move' on line.
![Selection_002](https://github.com/OCA/account-financial-tools/assets/20896369/2d436eba-d7cb-401f-a333-23f4f872646a)
4. Error because It creates a new move and **a new template** (i test by comment error. it will create new template and move_type is entry)

astirpe commented:
@Saran440 I just ported your latest fix to V16 a6dae3b but with context passed in `def create_move()`. It's better to pass the context within `def create_move()` so that it also fixes the "Create All Moves" button (that calls the same method).

Saran440 commented:
@astirpe Fixed. I change it to pass context in function `def create_move()`. Thank you for reviewed.

ps-tubtim reviewed (approved):
Functional Test: Ok

kittiu commented:
/ocabot merge nobump
